server/models/visit.js

```
@@ -0,0 +1,157 @@
+      '  RETURNING id ' +
+      ') SELECT id FROM new_page ' +
+      'UNION SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4';
+    var lazyCreateParams = [uuid.v4(), userId, url, urlHash, title, new Date().toJSON()];
+    var createVisitQuery = 'INSERT INTO visits ' +
+      '(id, user_id, user_page_id, visited_at, updated_at) ' +
+      'VALUES ($1, $2, $3, $4, $4)';
+    var userPageId;
+
+    // try to insert into pg; handle pg errors; insert into es; then we're done.
+    // don't pass results to fulfillment callback, because they aren't used by the caller.
+    postgres.query(lazyCreateUserPageQuery, lazyCreateParams)
+      .fail(visit._onRejected.bind(visit, name + ' failed', cb))
+      .then(function(results) {
+        // TODO will userPageId be correct? find out :-P
+        userPageId = results && results.id;
+        var visitParams = [visitId, userId, userPageId, visitedAt];
+        // return the new query's promise to avoid nesting promise chains
+        return postgres.query(createVisitQuery, visitParams);
+      })
+      .fail(visit._onRejected.bind(visit, name + ' postgres insert failed', cb))
+      .then(function() {
+        var esQuery = {
+          index: 'chronicle',
+          type: 'userPages',
+          id: userPageId,
+          body: {
+            userId: userId,
+            id: userPageId,
+            url: url,
+            urlHash: urlHash,
+            title: title,
+            visitedAt: visitedAt
+          }
+        };
+        // if the scraper elasticsearch task happens first, this one will abort, to avoid
+        // overwriting all the scraped data with nothing
+        return elasticsearch.query('create', esQuery)
+          .then(null, function(err) {
+            // TODO if the error is due to the thing already existing, it's not a problem...
+            // swallowing errors for now, until I look up how to catch that specific error
```

this is a bad place to swallow all errors. fix.